### PR TITLE
Adds typings for rusha

### PR DESCRIPTION
**Pull request by jacoscaz**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test rusha`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Adding a new definition:

- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [X] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [X] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

More information:

[rusha](https://www.npmjs.com/package/rusha) is a fast, pure-js implementation of the SHA1 hashing algorithm.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## rusha ([<kbd>unpkg</kbd>](https://unpkg.com/browse/rusha@latest/))
was missing the following properties:
1. The declaration doesn't match the JavaScript module 'rusha'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.
1. toHex
1. ceilHeapSize
1. isDedicatedWorkerScope
1. Rusha
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9a2b171dd7c9f841fc253821af4abb7dee943c3e
</p>

**Review by peterblazejewicz — COMMENTED** [review]
Shouldn't this export regular class with constructor `@deprecated`?
https://github.com/srijs/rusha#normal-usage

**Comment by jacoscaz** [comment]
@peterblazejewicz thank you for having a look at this. Developing typings for the deprecated API seems somewhat counter-productive to me as it encourages further use of that API. However, this is my first PR to DefinitelyTyped - if you believe it's better to do so I will continue working on this to cover the deprecated class.

**Review by rubensworks — APPROVED** [review]
Looks good to me!

File: types/rusha/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    id: string;
+    data: string|number[]|ArrayBuffer|Buffer;
+}
+
+interface WorkerResponse {
+    id: string;
+    hash: string;
+}
+
+interface Worker {
```
Comment (srijs) [inline types/rusha/index.d.ts:24]: Wondering if this could/should inherit from the built-in `Worker` type?

Comment (jacoscaz) [reply types/rusha/index.d.ts:24]: Sounds good. I’ll make the change.

File: types/rusha/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        "strictFunctionTypes": true,
+        "strictNullChecks": true,
+        "baseUrl": "../",
+        "typeRoots": [
+            "../"
+        ],
+        "types": [],
+        "noEmit": true,
+        "forceConsistentCasingInFileNames": true,
+        "esModuleInterop": true
```
Comment (peterblazejewicz) [inline types/rusha/tsconfig.json:18]:
```suggestion
        "forceConsistentCasingInFileNames": true
```
this is old, plain CommonJS module, no export default, etc. The flag is usualy for consumers of the DT package, not used here.
see also comment for test

File: types/rusha/rusha-tests.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import rusha from 'rusha';
```
Comment (peterblazejewicz) [inline types/rusha/rusha-tests.ts:1]:
```suggestion
import Rusha = require('rusha');
```
Notice `Rusha`, be consistent with documented conventions, this is a class-like object

File: types/rusha/rusha-tests.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+rusha; // $$ExpectType Rusha
```
Comment (peterblazejewicz) [inline types/rusha/rusha-tests.ts:3]:
```suggestion
Rusha; // $$ExpectType Rusha
```

File: types/rusha/rusha-tests.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+const hash = rusha.createHash(); // $$ExpectType Hash
```
Comment (peterblazejewicz) [inline types/rusha/rusha-tests.ts:5]:
```suggestion
const hash = Rusha.createHash(); // $$ExpectType Hash
```

File: types/rusha/rusha-tests.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+hash.update('Hello, World'); // // $ExpectType Hash
+hash.digest('hex'); // $ExpectType string
+hash.digest(); // $ExpectType
+
+const worker = rusha.createWorker(); // $$ExpectType Worker
```
Comment (peterblazejewicz) [inline types/rusha/rusha-tests.ts:10]:
```suggestion
const worker = Rusha.createWorker(); // $$ExpectType Worker
```

**Review by peterblazejewicz — COMMENTED** [review]
LGMT!
@jacoscaz

**Review by peterblazejewicz — APPROVED** [review]

**Comment by peterblazejewicz** [comment]
> @peterblazejewicz thank you for having a look at this. Developing typings for the deprecated API seems somewhat counter-productive to me as it encourages further use of that API. However, this is my first PR to DefinitelyTyped - if you believe it's better to do so I will continue working on this to cover the deprecated class.

I agree, I'd only consider if this change would be better reflecting the nature of the 'Rusha', e.g.:

```diff
-interface Rusha {
-    createHash(): Hash;
-    createWorker(): Worker;
-    disableWorkerBehaviour(): void;
+declare abstract class Rusha {
+    static createHash(): Hash;
+    static createWorker(): Worker;
+    static disableWorkerBehaviour(): void;
 }

-declare const Rusha: Rusha;
-
 export = Rusha;
```
this should fix the issue with the warning you're seeing, prevent usage via 'new' (as inteneted). The usage of 'abstract' here is perfectly valid, and you could consider using it, if it make sense. Just 2c

File: types/rusha/index.d.ts (hunk 1 above)
Comment (jacoscaz) [reply types/rusha/index.d.ts]: Done!

File: types/rusha/rusha-tests.ts (hunk 3 above)
Comment (jacoscaz) [reply types/rusha/rusha-tests.ts]: Done!

File: types/rusha/rusha-tests.ts (hunk 5 above)
Comment (jacoscaz) [reply types/rusha/rusha-tests.ts]: Done!

File: types/rusha/tsconfig.json (hunk 2 above)
Comment (jacoscaz) [reply types/rusha/tsconfig.json]: Done!

**Comment by jacoscaz** [comment]
@srijs and @peterblazejewicz - thank you for your guidance. I've incorporated all of your suggestions minus the very last one by @peterblazejewicz as I had already tried using the abstract class but ended up with the following:

```
ERROR: 31:18  no-unnecessary-class  Every member of this class is static. Use namespaces or plain objects instead.
```

File: types/rusha/index.d.ts [hunk 7]
```diff
@@ -2,34 +2,35 @@
 // Project: https://github.com/srijs/rusha#readme
 // Definitions by: Jacopo Scazzosi <https://github.com/jacoscaz>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 4.0
 
 /// <reference types="node" />
 
-interface Hash {
-    update(value: string|number[]|ArrayBuffer|Buffer): Hash;
+interface RushaHash {
```
Comment (peterblazejewicz) [inline types/rusha/index.d.ts:9]: this is a Hash, why renaming existing symbols? (think maintaince)
https://github.com/srijs/rusha/blob/master/src/hash.js#L6

**Review by peterblazejewicz — APPROVED** [review]

File: types/rusha/index.d.ts (hunk 7 above)
Comment (jacoscaz) [reply types/rusha/index.d.ts:9]: What is the best practice here? I had to rename `Worker` into `RushaRowker` as, with the addition of the `DOM` library, it had started shadowing `Worker`. I thought renaming `Hash` into something more specific would make things more uniform.

Comment (peterblazejewicz) [reply types/rusha/index.d.ts:9]: as it is late here, I would say, we're describing existing bestiary, so we keep existing names. When we produce brand new beast, we give it new name. Applies to types and objects as well as to beasts. 
Just think about users of your your definition, they would not expect `RushaHash`, it does not existisn in the package readme. THey would expect 'Hash'

Comment (jacoscaz) [reply types/rusha/index.d.ts]: Done, renamed `RushaHash` back to `Hash`

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by jacoscaz** [comment]
Ready to merge

**Pull request merged**